# Sesión 2 - Data Wrangling y Pivots

## Antes de empezar

La tabla `f1_personas` ya debe estar descargada en `Materiales/Insumos/ENDI/BDD_ENDI_R2_rds/`.

**Pregunta que guía la sesión:**  
Tengo los datos en memoria. ¿Cómo los preparo para responder una pregunta sobre desnutrición infantil?

## Importaciones

In [115]:
import pyreadr
import polars as pl
import polars.selectors as cs
import pandas as pd

## Carga y primera vista

Cargamos `f1_personas` igual que en la sesión 1.

In [116]:
ruta = "../../Materiales/Insumos/ENDI/BDD_ENDI_R2_rds/BDD_ENDI_R2_f1_personas.rds"

df = pl.from_pandas(pyreadr.read_r(ruta)[None])
df.shape

(93242, 124)

In [117]:
# los tipos incorrectos se asignan desde el read_r
(pyreadr.read_r(ruta)[None]).info()

# (pyreadr.read_r(ruta)[None]).dtypes

<class 'pandas.DataFrame'>
RangeIndex: 93242 entries, 0 to 93241
Columns: 124 entries, id_upm to ane6_23_new
dtypes: float64(38), int32(15), object(61), str(10)
memory usage: 89.2+ MB


### Revisión del schema

Antes de cualquier operación, revisamos qué tipos infirió Polars. Algunos serán incorrectos.

In [118]:
# Schema completo
df.schema

Schema([('id_upm', String),
        ('id_viv', String),
        ('id_hogar', String),
        ('id_per', String),
        ('id_mef', String),
        ('fecha_anio', String),
        ('fecha_mes', String),
        ('fecha_dia', String),
        ('fexp', Float64),
        ('estrato', String),
        ('area', Float64),
        ('region', Float64),
        ('prov', Float64),
        ('parr_pri', Float64),
        ('etnia', Float64),
        ('persona', String),
        ('altitud', Int32),
        ('edaddias', Float64),
        ('grupo_edad_nin', Float64),
        ('nivins_mef', Float64),
        ('f1_s1_1', Int32),
        ('f1_s1_2', Int32),
        ('f1_s1_3_1', Int32),
        ('f1_s1_4_1', Int32),
        ('f1_s1_4_2', Int32),
        ('f1_s1_4_3', Int32),
        ('f1_s1_5', Int32),
        ('f1_s1_6', Int32),
        ('f1_s1_6_2', Int64),
        ('f1_s1_7_a', Int64),
        ('f1_s1_7_b', Int64),
        ('f1_s1_7_c', Int64),
        ('f1_s1_7_d', Int64),
        ('f1_s1_7_e', Int6

In [119]:
df.head(5)

id_upm,id_viv,id_hogar,id_per,id_mef,fecha_anio,fecha_mes,fecha_dia,fexp,estrato,area,region,prov,parr_pri,etnia,persona,altitud,edaddias,grupo_edad_nin,nivins_mef,f1_s1_1,f1_s1_2,f1_s1_3_1,f1_s1_4_1,f1_s1_4_2,f1_s1_4_3,f1_s1_5,f1_s1_6,f1_s1_6_2,f1_s1_7_a,f1_s1_7_b,f1_s1_7_c,f1_s1_7_d,f1_s1_7_e,f1_s1_7_f,f1_s1_8,f1_s1_9,…,f1_s5_3_2,f1_s5_3_3,f1_s5_4_1,f1_s5_4_2,f1_s5_4_3,f1_s5_5_1,f1_s5_5_2,f1_s5_5_3,f1_s5_6_1,f1_s5_6_2,f1_s5_6_3,f1_s5_7,f1_s6_1,f1_s6_2,f1_s6_3,f1_s6_4_1,f1_s6_4_2,f1_s6_5_1,f1_s6_5_2,f1_s6_5_3,f1_s6_6,quintil,pobreza,nbi_1,dcronica_2,dglobal_2,daguda_2,dcronica,dglobal,daguda,dcronica2_5,dglobal2_5,daguda2_5,ane6_59,ane6_59_new,ane6_23,ane6_23_new
str,str,str,str,str,str,str,str,f64,str,f64,f64,f64,f64,f64,str,i32,f64,f64,f64,i32,i32,i32,i32,i32,i32,i32,i32,i64,i64,i64,i64,i64,i64,i64,i32,i64,…,i64,i64,f64,f64,f64,f64,f64,f64,f64,f64,f64,i64,i64,i64,f64,i64,i64,i64,i64,i64,i64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010101""",null,"""2024""","""05""","""15""",0.297295,"""2712""",1.0,1.0,1.0,2.0,4.0,"""01""",2609,24473.0,null,null,1,1,67,14,5,1957,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010102""",null,"""2024""","""05""","""15""",0.297295,"""2712""",1.0,1.0,1.0,2.0,4.0,"""02""",2609,23107.0,null,null,2,2,63,8,2,1961,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010103""",null,"""2024""","""05""","""15""",0.297295,"""2712""",1.0,1.0,1.0,2.0,4.0,"""03""",2609,14719.0,null,null,3,1,40,27,1,1984,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010104""","""0101500088010104""","""2024""","""05""","""15""",0.297295,"""2712""",1.0,1.0,1.0,2.0,4.0,"""04""",2609,11903.0,null,2.0,5,2,32,13,10,1991,1,1,null,1,1,1,1,1,1,2,null,…,5,2024,60.8,60.8,null,null,null,null,150.4,150.4,null,null,4,1,16.0,12,34,15,5,2024,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010105""",null,"""2024""","""05""","""15""",0.297295,"""2712""",1.0,1.0,1.0,2.0,4.0,"""05""",2609,1030.0,4.0,null,6,1,2,20,7,2021,1,1,null,null,null,null,null,null,null,2,null,…,5,2024,12.7,12.7,null,null,null,null,88.0,88.0,null,null,4,1,13.3,12,31,15,5,2024,null,3.0,0.0,0.0,null,null,null,0.0,0.0,0.0,0.0,0.0,0.0,4.0,0.0,null,null


In [120]:
# Cuántas columnas hay por tipo
from collections import Counter

conteo = Counter(str(t) for t in df.schema.dtypes())
pl.DataFrame(conteo)

String,Float64,Int32,Int64
i64,i64,i64,i64
10,38,15,61


## Corrección de tipos y recodificación

### ¿Qué está mal?

Polars infiere los tipos columna por columna. Lo hace bien para variables genuinamente numéricas como `fexp` o `talla`, pero se equivoca con columnas que contienen **códigos numéricos**: números que representan categorías.

Un `Float64` donde debería haber una categoría no genera error: simplemente devuelve resultados sin sentido. `df["area"].mean()` da 1.38. Nadie lo advierte.

In [121]:
# Ver los tipos actuales de las variables que nos interesan
df.select(["area", "region", "prov", "etnia", "f1_s1_2","estrato", "parr_pri", "grupo_edad_nin"]).schema

Schema([('area', Float64),
        ('region', Float64),
        ('prov', Float64),
        ('etnia', Float64),
        ('f1_s1_2', Int32),
        ('estrato', String),
        ('parr_pri', Float64),
        ('grupo_edad_nin', Float64)])

In [122]:
# df

### Recodificación con `pl.when().then().otherwise()`

Mapeamos los códigos numéricos a etiquetas legibles usando los valores del diccionario de variables de la ENDI. Al mismo tiempo declaramos el tipo correcto: `pl.Categorical` para variables nominales, `pl.Enum` para variables con orden natural.

In [123]:
df = df.with_columns([

    # area: nominal, sin orden entre categorías
    pl.when(pl.col("area") == 1).then(pl.lit("Urbano"))
      .when(pl.col("area") == 2).then(pl.lit("Rural"))
      .otherwise(pl.lit(""))
      .cast(pl.Categorical)
      .alias("area"),

    # region: nominal, sin orden geográfico implícito
    pl.when(pl.col("region") == 1).then(pl.lit("Sierra"))
      .when(pl.col("region") == 2).then(pl.lit("Costa"))
      .when(pl.col("region") == 3).then(pl.lit("Amazonia"))
      .when(pl.col("region") == 4).then(pl.lit("Insular"))
      .otherwise(pl.lit(None))
      .cast(pl.Categorical)
      .alias("region"),

    # f1_s1_2 (sexo): nominal
    pl.when(pl.col("f1_s1_2") == 1).then(pl.lit("Hombre"))
      .when(pl.col("f1_s1_2") == 2).then(pl.lit("Mujer"))
      .otherwise(pl.lit(None))
      .cast(pl.Categorical)
      .alias("f1_s1_2"),

    # etnia: nominal
    pl.when(pl.col("etnia") == 1).then(pl.lit("Indígena"))
      .when(pl.col("etnia") == 2).then(pl.lit("Afroecuatoriana/o"))
      .when(pl.col("etnia") == 3).then(pl.lit("Montubia/o"))
      .when(pl.col("etnia") == 4).then(pl.lit("Mestiza/o"))
      .when(pl.col("etnia") == 5).then(pl.lit("Blanca/o"))
      .when(pl.col("etnia") == 6).then(pl.lit("Otra"))
      .otherwise(pl.lit(None))
      .cast(pl.Categorical)
      .alias("etnia"),

    # prov: código string
    pl.col("prov").cast(pl.Int8).cast(pl.String).str.pad_start(2,"0"),

    # estrato y parr_pri: categóricas de diseño muestral
    pl.col("parr_pri").cast(pl.Int8),

])



In [124]:
# Renombrar f1_s1_2 -> sexo ya queda implícito en el alias de arriba
# Verificar
df.select(["area", "region", "f1_s1_2", "etnia","prov"]).head(5)

area,region,f1_s1_2,etnia,prov
cat,cat,cat,cat,str
"""Urbano""","""Sierra""","""Hombre""","""Mestiza/o""","""01"""
"""Urbano""","""Sierra""","""Mujer""","""Mestiza/o""","""01"""
"""Urbano""","""Sierra""","""Hombre""","""Mestiza/o""","""01"""
"""Urbano""","""Sierra""","""Mujer""","""Mestiza/o""","""01"""
"""Urbano""","""Sierra""","""Hombre""","""Mestiza/o""","""01"""


### Variables ordinales con `pl.Enum`

`grupo_edad_nin` tiene un orden natural: los grupos de edad van de menor a mayor. Con `pl.Enum` declaramos ese orden explícitamente. Polars lo usará al ordenar y en gráficos sin que tengamos que especificarlo cada vez.

In [125]:
orden_grupos = [
    "0-5 meses",
    "6-11 meses",
    "12-23 meses",
    "24-35 meses",
    "36-47 meses",
    "48-59 meses",
]

df.with_columns(
    pl.when(pl.col("grupo_edad_nin") == 1).then(pl.lit(orden_grupos[0]))
      .when(pl.col("grupo_edad_nin") == 5).then(pl.lit(orden_grupos[4]))
      .when(pl.col("grupo_edad_nin") == 3).then(pl.lit(orden_grupos[2]))
      .when(pl.col("grupo_edad_nin") == 4).then(pl.lit(orden_grupos[3]))
      .when(pl.col("grupo_edad_nin") == 2).then(pl.lit(orden_grupos[1]))
      .when(pl.col("grupo_edad_nin") == 6).then(pl.lit(orden_grupos[5]))
      .otherwise(pl.lit(None))
      .cast(pl.Enum(orden_grupos))
      .alias("grupo_edad_nin")
)
# ["grupo_edad_nin"].unique()

id_upm,id_viv,id_hogar,id_per,id_mef,fecha_anio,fecha_mes,fecha_dia,fexp,estrato,area,region,prov,parr_pri,etnia,persona,altitud,edaddias,grupo_edad_nin,nivins_mef,f1_s1_1,f1_s1_2,f1_s1_3_1,f1_s1_4_1,f1_s1_4_2,f1_s1_4_3,f1_s1_5,f1_s1_6,f1_s1_6_2,f1_s1_7_a,f1_s1_7_b,f1_s1_7_c,f1_s1_7_d,f1_s1_7_e,f1_s1_7_f,f1_s1_8,f1_s1_9,…,f1_s5_3_2,f1_s5_3_3,f1_s5_4_1,f1_s5_4_2,f1_s5_4_3,f1_s5_5_1,f1_s5_5_2,f1_s5_5_3,f1_s5_6_1,f1_s5_6_2,f1_s5_6_3,f1_s5_7,f1_s6_1,f1_s6_2,f1_s6_3,f1_s6_4_1,f1_s6_4_2,f1_s6_5_1,f1_s6_5_2,f1_s6_5_3,f1_s6_6,quintil,pobreza,nbi_1,dcronica_2,dglobal_2,daguda_2,dcronica,dglobal,daguda,dcronica2_5,dglobal2_5,daguda2_5,ane6_59,ane6_59_new,ane6_23,ane6_23_new
str,str,str,str,str,str,str,str,f64,str,cat,cat,str,i8,cat,str,i32,f64,enum,f64,i32,cat,i32,i32,i32,i32,i32,i32,i64,i64,i64,i64,i64,i64,i64,i32,i64,…,i64,i64,f64,f64,f64,f64,f64,f64,f64,f64,f64,i64,i64,i64,f64,i64,i64,i64,i64,i64,i64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64,f64
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010101""",null,"""2024""","""05""","""15""",0.297295,"""2712""","""Urbano""","""Sierra""","""01""",2,"""Mestiza/o""","""01""",2609,24473.0,null,null,1,"""Hombre""",67,14,5,1957,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010102""",null,"""2024""","""05""","""15""",0.297295,"""2712""","""Urbano""","""Sierra""","""01""",2,"""Mestiza/o""","""02""",2609,23107.0,null,null,2,"""Mujer""",63,8,2,1961,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010103""",null,"""2024""","""05""","""15""",0.297295,"""2712""","""Urbano""","""Sierra""","""01""",2,"""Mestiza/o""","""03""",2609,14719.0,null,null,3,"""Hombre""",40,27,1,1984,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010104""","""0101500088010104""","""2024""","""05""","""15""",0.297295,"""2712""","""Urbano""","""Sierra""","""01""",2,"""Mestiza/o""","""04""",2609,11903.0,null,2.0,5,"""Mujer""",32,13,10,1991,1,1,null,1,1,1,1,1,1,2,null,…,5,2024,60.8,60.8,null,null,null,null,150.4,150.4,null,null,4,1,16.0,12,34,15,5,2024,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""0101500088""","""010150008801""","""01015000880101""","""0101500088010105""",null,"""2024""","""05""","""15""",0.297295,"""2712""","""Urbano""","""Sierra""","""01""",2,"""Mestiza/o""","""05""",2609,1030.0,"""24-35 meses""",null,6,"""Hombre""",2,20,7,2021,1,1,null,null,null,null,null,null,null,2,null,…,5,2024,12.7,12.7,null,null,null,null,88.0,88.0,null,null,4,1,13.3,12,31,15,5,2024,null,3.0,0.0,0.0,null,null,null,0.0,0.0,0.0,0.0,0.0,0.0,4.0,0.0,null,null
…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…,…
"""2402500090""","""240250009008""","""24025000900801""","""2402500090080105""",null,"""2024""","""06""","""30""",0.979144,"""2411""","""Urbano""","""Costa""","""24""",1,"""Mestiza/o""","""05""",10,6427.0,null,null,3,"""Hombre""",17,25,11,2006,1,1,null,1,1,1,1,1,1,2,null,…,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,3.0,0.0,0.0,null,null,null,null,null,null,null,null,null,null,null,null,null
"""2402500090""","""240250009008""","""24025000900801""","""2402500090080106""","""2402500090080106""","""2024""","""06""","""

O de forma programática si son demasiado largas las listas. (!no necesito el alias cuando empieza con pl.col y no quiewro una nueva columna¡)

In [126]:
df["grupo_edad_nin"].value_counts()

grupo_edad_nin,count
f64,u32
null,70055
6.0,5113
4.0,4563
5.0,4898
2.0,2301
3.0,4595
1.0,1717


In [127]:
mapeo = {i+1: v for i, v in enumerate(orden_grupos)}

In [128]:
df.with_columns(
    pl.col("grupo_edad_nin")
    .cast(pl.Int8)
    .cast(pl.String)
    .replace(mapeo)
    .cast(pl.Enum(orden_grupos))
)["grupo_edad_nin"].value_counts().sort("grupo_edad_nin")

grupo_edad_nin,count
enum,u32
null,70055
"""0-5 meses""",1717
"""6-11 meses""",2301
"""12-23 meses""",4595
"""24-35 meses""",4563
"""36-47 meses""",4898
"""48-59 meses""",5113


In [129]:
df = df.with_columns(
        pl.col("grupo_edad_nin")
        .cast(pl.Int8)
        .cast(pl.String)
        .replace(mapeo)
        .cast(pl.Enum(orden_grupos))
)

## Renombrar y seleccionar columnas

### Buena práctica: nombres autoexplicativos

Los nombres originales de la ENDI siguen la convención del formulario: `f1_s5_2_1` significa formulario 1, sección 5, pregunta 2, parte 1. Son precisos pero crípticos. Antes de trabajar con las columnas que nos interesan, las renombramos.

Un buen nombre de columna responde la pregunta: *¿qué contiene este campo?* sin necesidad de consultar el diccionario.

In [130]:
df = df.rename({
    # Identificadores (ya son claros, pero los dejamos explícitos)
    "f1_s1_2": "sexo",

    # Fecha de nacimiento del niño/a (sección 5, pregunta 2)
    "f1_s5_2_1": "nac_dia",
    "f1_s5_2_2": "nac_mes",
    "f1_s5_2_3": "nac_anio",

    # Fecha de la medición (sección 5, pregunta 3)
    "f1_s5_3_1": "med_dia",
    "f1_s5_3_2": "med_mes",
    "f1_s5_3_3": "med_anio",

    # Tomas de longitud (cm, menores de 2 años, acostados)
    "f1_s5_5_1": "long_1",
    "f1_s5_5_2": "long_2",
    "f1_s5_5_3": "long_3",

    # Tomas de talla (cm, 2 años y más, de pie)
    "f1_s5_6_1": "talla_1",
    "f1_s5_6_2": "talla_2",
    "f1_s5_6_3": "talla_3",
})

# Verificar que los nombres nuevos están en el schema
print(df.select(["sexo", "nac_dia", "nac_mes", "nac_anio",
          "med_dia", "med_mes", "med_anio",
          "long_1", "long_2", "long_3",
          "talla_1", "talla_2", "talla_3"]).schema)

Schema({'sexo': Categorical, 'nac_dia': Int64, 'nac_mes': Int64, 'nac_anio': Int64, 'med_dia': Int64, 'med_mes': Int64, 'med_anio': Int64, 'long_1': Float64, 'long_2': Float64, 'long_3': Float64, 'talla_1': Float64, 'talla_2': Float64, 'talla_3': Float64})


### Reducir el dataframe

Tenemos 124 columnas y utilizamos muchas menos. Nos quedamos solo con las que necesitamos para el indicador de desnutrición crónica.

**Regla práctica:** seleccionar primero, filtrar, calcular después. Un dataframe más pequeño es más rápido y más fácil de inspeccionar.

In [132]:
cols_interes = ["id_per", "id_hogar", "id_upm", "estrato", "fexp",
    "area", "region", "prov", "sexo",

    # Variables calculadas (ya vienen en la tabla)
    "edaddias", "grupo_edad_nin",
    "dcronica", "dcronica2_5",

    # Fechas y mediciones antropométricas (para recalcular desde cero)
    "nac_dia", "nac_mes", "nac_anio",
    "med_dia", "med_mes", "med_anio",
    "long_1", "long_2", "long_3",
    "talla_1", "talla_2", "talla_3",
    "f1_s1_3_1"
]

df = df.select(cols_interes)

print(f"Columnas después del select: {df.shape[1]}")
print(df.schema)

Columnas después del select: 26
Schema({'id_per': String, 'id_hogar': String, 'id_upm': String, 'estrato': String, 'fexp': Float64, 'area': Categorical, 'region': Categorical, 'prov': String, 'sexo': Categorical, 'edaddias': Float64, 'grupo_edad_nin': Enum(categories=['0-5 meses', '6-11 meses', '12-23 meses', '24-35 meses', '36-47 meses', '48-59 meses']), 'dcronica': Float64, 'dcronica2_5': Float64, 'nac_dia': Int64, 'nac_mes': Int64, 'nac_anio': Int64, 'med_dia': Int64, 'med_mes': Int64, 'med_anio': Int64, 'long_1': Float64, 'long_2': Float64, 'long_3': Float64, 'talla_1': Float64, 'talla_2': Float64, 'talla_3': Float64, 'f1_s1_3_1': Int32})


### Filtrar los casos con medición antropométrica

No todos los miembros del hogar tienen medición de talla y peso: solo los menores de 5 años. Filtramos para quedarnos con el subconjunto relevante.

El campo `grupo_edad_nin` ya es un indicador de si el registro tiene datos antropométricos: viene como nulo para adultos.

In [133]:
# Menores de 5 años con medición (tienen grupo_edad_nin no nulo)
df_nin = df.filter(pl.col("grupo_edad_nin").is_not_null())

print(f"Filas totales:          {df.shape[0]:>7,}")
print(f"Menores con medición:   {df_nin.shape[0]:>7,}")
print(f"Proporción del total:   {df_nin.shape[0]/df.shape[0]:.1%}")

Filas totales:           93,242
Menores con medición:    23,187
Proporción del total:   24.9%


In [135]:
df.filter(pl.col("f1_s1_3_1")<5).equals(df.filter(pl.col("grupo_edad_nin").is_not_null()))

True

##  Calcular edad en días desde las fechas crudas

### El problema

La variable `edaddias` ya viene calculada en la tabla. Sin embargo, vamos a recalcularla desde las columnas crudas por dos razones:

1. Para entender exactamente cómo se construye (el peso del denominador en la fórmula del indicador depende de esto)
2. Para aprender a trabajar con fechas en Polars, que es necesario en el indicador de vacunación de rotavirus

La lógica: edad en días = fecha de medición - fecha de nacimiento.

### Construir fechas desde columnas separadas

La ENDI guarda día, mes y año en columnas separadas. Hay que combinarlas en una fecha.

In [74]:
# Revisar los tipos de las columnas de fecha antes de operar
print(df_nin.select(["nac_dia", "nac_mes", "nac_anio",
                     "med_dia", "med_mes", "med_anio"]).schema)

Schema({'nac_dia': Int64, 'nac_mes': Int64, 'nac_anio': Int64, 'med_dia': Int64, 'med_mes': Int64, 'med_anio': Int64})


In [75]:
# Construir las fechas
# pl.date(year, month, day) crea una columna de tipo Date a partir de componentes enteros
df_nin = df_nin.with_columns([
    pl.date(
        pl.col("nac_anio").cast(pl.Int32),
        pl.col("nac_mes").cast(pl.Int32),
        pl.col("nac_dia").cast(pl.Int32),
    ).alias("fecha_nac"),

    pl.date(
        pl.col("med_anio").cast(pl.Int32),
        pl.col("med_mes").cast(pl.Int32),
        pl.col("med_dia").cast(pl.Int32),
    ).alias("fecha_med"),
])

print(df_nin.select(["fecha_nac", "fecha_med"]).head(5))

shape: (5, 2)
┌────────────┬────────────┐
│ fecha_nac  ┆ fecha_med  │
│ ---        ┆ ---        │
│ date       ┆ date       │
╞════════════╪════════════╡
│ 2021-07-20 ┆ 2024-05-15 │
│ 2023-10-03 ┆ 2024-05-15 │
│ 2019-12-08 ┆ 2024-05-06 │
│ 2021-07-19 ┆ 2024-05-06 │
│ 2020-07-11 ┆ 2024-05-06 │
└────────────┴────────────┘


### 4.3 Calcular edad en días

La diferencia entre dos fechas en Polars devuelve una duración. Con `.dt.total_days()` la convertimos a un entero.

In [76]:
df_nin = df_nin.with_columns(
    (pl.col("fecha_med") - pl.col("fecha_nac"))
    .dt.total_days()
    .alias("edad_dias_calc")
)

# Comparar con la variable precalculada del INEC
print(
    df_nin.select([
        "edaddias",
        "edad_dias_calc",
        (pl.col("edaddias") - pl.col("edad_dias_calc")).abs().alias("diferencia")
    ]).describe()
)

shape: (9, 4)
┌────────────┬────────────┬────────────────┬────────────┐
│ statistic  ┆ edaddias   ┆ edad_dias_calc ┆ diferencia │
│ ---        ┆ ---        ┆ ---            ┆ ---        │
│ str        ┆ f64        ┆ f64            ┆ f64        │
╞════════════╪════════════╪════════════════╪════════════╡
│ count      ┆ 23187.0    ┆ 22340.0        ┆ 22340.0    │
│ null_count ┆ 0.0        ┆ 847.0          ┆ 847.0      │
│ mean       ┆ 956.732393 ┆ 957.82744      ┆ 0.106088   │
│ std        ┆ 518.324098 ┆ 519.205911     ┆ 0.879879   │
│ min        ┆ 1.0        ┆ 1.0            ┆ 0.0        │
│ 25%        ┆ 509.0      ┆ 507.0          ┆ 0.0        │
│ 50%        ┆ 966.0      ┆ 969.0          ┆ 0.0        │
│ 75%        ┆ 1413.0     ┆ 1417.0         ┆ 0.0        │
│ max        ┆ 1825.0     ┆ 1825.0         ┆ 52.0       │
└────────────┴────────────┴────────────────┴────────────┘


In [77]:
# ¿Hay casos donde difieren?
distintos = df_nin.filter(
    (pl.col("edaddias") - pl.col("edad_dias_calc")).abs() > 1
)
print(f"Casos con diferencia > 1 día: {distintos.shape[0]}")

Casos con diferencia > 1 día: 429


##  Joins

### Tipos de join

| Tipo | ¿Qué conserva? | Cuándo usarlo |
|---|---|---|
| `left` | Todas las filas de la tabla izquierda | Agregar columnas de otra tabla sin perder casos |
| `right` | Todas las filas de la tabla derecha | Menos común; equivale a un `left` con las tablas invertidas |
| `inner` | Solo filas con coincidencia en ambas tablas | Cuando se quiere el subconjunto que existe en las dos |
| `full` | Todas las filas de ambas tablas | Cuando no se quiere perder ningún caso de ninguna tabla |
| `semi` | Filas de la izquierda que tienen coincidencia, sin agregar columnas | Filtrar sin expandir |
| `anti` | Filas de la izquierda que **no** tienen coincidencia | Encontrar casos que faltan en la otra tabla |

### Left join: personas con hogar

El caso más común en la ENDI: queremos agregar variables del hogar a la tabla de personas. Cada persona pertenece a un hogar (`id_hogar`), y queremos conservar todas las personas aunque algún hogar no tenga datos.

In [78]:
# Cargar f1_hogar
ruta_hogar = "../../Materiales/Insumos/ENDI/BDD_ENDI_R2_rds/BDD_ENDI_R2_f1_hogar.rds"

df_hogar = pl.from_pandas(
    pyreadr.read_r(ruta_hogar)[None]
)

print(f"Hogares: {df_hogar.shape[0]:,} filas  |  {df_hogar.shape[1]} columnas")

Hogares: 20,186 filas  |  133 columnas


In [79]:
# Seleccionar solo las columnas de hogar que nos interesan
# f1_s3_2: tipo de vivienda
# f1_s7_1: fuente de agua para beber
df_hogar_min = df_hogar.select(["id_hogar", "f1_s3_2", "f1_s7_1"])

df_hogar_min = df_hogar_min.rename({
    "f1_s3_2":  "tipo_vivienda",
    "f1_s7_1": "fuente_agua",
})

print(df_hogar_min.head(3))

shape: (3, 3)
┌────────────────┬───────────────┬─────────────┐
│ id_hogar       ┆ tipo_vivienda ┆ fuente_agua │
│ ---            ┆ ---           ┆ ---         │
│ str            ┆ i32           ┆ i32         │
╞════════════════╪═══════════════╪═════════════╡
│ 01115000040201 ┆ 1             ┆ 1           │
│ 01115000040401 ┆ 1             ┆ 1           │
│ 01115000040501 ┆ 1             ┆ 1           │
└────────────────┴───────────────┴─────────────┘


In [80]:
# Left join: agregar variables del hogar a los niños
df_nin_hogar = df_nin.join(df_hogar_min,on="id_hogar",how="left")

print(f"Filas antes del join: {df_nin.shape[0]:,}")
print(f"Filas después:        {df_nin_hogar.shape[0]:,}")
print(f"\nNuevas columnas: {[c for c in df_nin_hogar.columns if c not in df_nin.columns]}")

Filas antes del join: 23,187
Filas después:        23,187

Nuevas columnas: ['tipo_vivienda', 'fuente_agua']


In [81]:
df_nin_hogar.head()

id_per,id_hogar,id_upm,estrato,fexp,area,region,prov,sexo,edaddias,grupo_edad_nin,dcronica,dcronica2_5,nac_dia,nac_mes,nac_anio,med_dia,med_mes,med_anio,long_1,long_2,long_3,talla_1,talla_2,talla_3,f1_s1_3_1,fecha_nac,fecha_med,edad_dias_calc,tipo_vivienda,fuente_agua
str,str,str,str,f64,cat,cat,str,cat,f64,enum,f64,f64,i64,i64,i64,i64,i64,i64,f64,f64,f64,f64,f64,f64,i32,date,date,i64,i32,i32
"""0101500088010105""","""01015000880101""","""0101500088""","""2712""",0.297295,"""Urbano""","""Sierra""","""01""","""Hombre""",1030.0,"""24-35 meses""",0.0,0.0,20,7,2021,15,5,2024,null,null,null,88.0,88.0,null,2,2021-07-20,2024-05-15,1030,2,1
"""0101500088010106""","""01015000880101""","""0101500088""","""2712""",0.297295,"""Urbano""","""Sierra""","""01""","""Mujer""",225.0,"""6-11 meses""",0.0,null,3,10,2023,15,5,2024,65.4,65.5,null,null,null,null,0,2023-10-03,2024-05-15,225,2,1
"""0101500088020104""","""01015000880201""","""0101500088""","""2712""",0.297295,"""Urbano""","""Sierra""","""01""","""Hombre""",1611.0,"""48-59 meses""",0.0,0.0,8,12,2019,6,5,2024,null,null,null,102.1,102.1,null,4,2019-12-08,2024-05-06,1611,2,1
"""0101500088040105""","""01015000880401""","""0101500088""","""2712""",0.297295,"""Urbano""","""Sierra""","""01""","""Hombre""",1022.0,"""24-35 meses""",1.0,1.0,19,7,2021,6,5,2024,null,null,null,86.3,86.1,null,2,2021-07-19,2024-05-06,1022,2,5
"""0101500088050103""","""01015000880501""","""0101500088""","""2712""",0.297295,"""Urbano""","""Sierra""","""01""","""Mujer""",1395.0,"""36-47 meses""",0.0,0.0,11,7,2020,6,5,2024,null,null,null,97.4,97.5,null,3,2020-07-11,2024-05-06,1395,2,1


In [82]:
# Verificar: ¿algún niño quedó sin datos de hogar?
sin_hogar = df_nin_hogar.filter(pl.col("tipo_vivienda").is_null()).shape[0]
print(f"Niños sin datos de hogar después del join: {sin_hogar}")

Niños sin datos de hogar después del join: 0


### Inner join vs left join - la diferencia

Para ver la diferencia concretamente, comparamos el número de filas que devuelve cada tipo.

In [83]:
n_left  = df_nin.join(df_hogar_min, on="id_hogar", how="left").shape[0]
n_inner = df_nin.join(df_hogar_min, on="id_hogar", how="inner").shape[0]
n_anti  = df_nin.join(df_hogar_min, on="id_hogar", how="anti").shape[0]

print(f"left join:  {n_left:>7,} filas  (conserva todos los niños)")
print(f"inner join: {n_inner:>7,} filas  (solo niños con hogar coincidente)")
print(f"anti join:  {n_anti:>7,} filas  (niños sin hogar coincidente)")
print(f"\nSuma inner + anti = {n_inner + n_anti:,}  (debe coincidir con el total de niños)")

left join:   23,187 filas  (conserva todos los niños)
inner join:  23,187 filas  (solo niños con hogar coincidente)
anti join:        0 filas  (niños sin hogar coincidente)

Suma inner + anti = 23,187  (debe coincidir con el total de niños)
